### PR TITLE
Issue #166 - Replica.updateMetastore calls metastore with batched partition lists.

### PR DESCRIPTION
**Pull request by barnharts4**
Fixes Issue #166 

Log file output from running test on table with 5019 partitions:
```
21:01:53.209 [main] INFO  com.hotels.bdp.circustrain.core.replica.Replica - Updating replica table metadata.
21:01:53.334 [main] INFO  com.hotels.hcommon.hive.metastore.client.closeable.CloseableMetaStoreClientInvocationHandler - Couldn't invoke method public abstract org.apache.hadoop.hive.metastore.api.Table org.apache.hadoop.hive.metastore.IMetaStoreClient.getTable(java.lang.String,java.lang.String) throws org.apache.hadoop.hive.metastore.api.MetaException,org.apache.thrift.TException,org.apache.hadoop.hive.metastore.api.NoSuchObjectException
21:11:02.672 [main] INFO  com.hotels.bdp.circustrain.core.replica.Replica - Creating 5019 new partitions.
21:11:02.677 [main] INFO  com.hotels.bdp.circustrain.core.replica.Replica - Creating partitions 0 through 999
21:15:21.485 [main] INFO  com.hotels.bdp.circustrain.core.replica.Replica - Creating partitions 1000 through 1999
21:19:35.692 [main] INFO  com.hotels.bdp.circustrain.core.replica.Replica - Creating partitions 2000 through 2999
21:23:57.927 [main] INFO  com.hotels.bdp.circustrain.core.replica.Replica - Creating partitions 3000 through 3999
21:28:13.187 [main] INFO  com.hotels.bdp.circustrain.core.replica.Replica - Creating partitions 4000 through 4999
21:32:30.142 [main] INFO  com.hotels.bdp.circustrain.core.replica.Replica - Creating partitions 5000 through 5018
21:32:35.818 [main] INFO  hive.metastore - Closed a connection to metastore, current connections: 0
21:32:35.818 [main] INFO  com.hotels.bdp.circustrain.core.PartitionedTableReplication - Replicated 5019 partitions of table test.barnparttest5000.
```

### CODE REVIEW

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java [hunk 1]
```diff
@@ -189,7 +222,13 @@ public void updateMetadata(
       if (!partitionsToAlter.isEmpty()) {
         LOG.info("Altering {} existing partitions.", partitionsToAlter.size());
         try {
-          client.alter_partitions(replicaDatabaseName, replicaTableName, partitionsToAlter);
+          int counter = 0;
+          for (List<Partition> sublist : Lists.partition(partitionsToAlter, partitionBatchSize)) {
+            int start = counter * partitionBatchSize;
+            LOG.info("Altering partitions {} through {}", start, start + sublist.size() - 1);
+            client.alter_partitions(replicaDatabaseName, replicaTableName, partitionsToAlter);
```
Comment (abhimanyugupta07) [inline circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java]: shouldn't this be `sublist` instead of `partitionsToAlter`.

`client.alter_partitions(replicaDatabaseName, replicaTableName, partitionsToAlter);`

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java [hunk 2]
```diff
@@ -203,7 +242,13 @@ public void updateMetadata(
       if (!statisticsToSet.isEmpty()) {
         LOG.info("Setting column statistics for {} partitions.", statisticsToSet.size());
         try {
-          client.setPartitionColumnStatistics(new SetPartitionsStatsRequest(statisticsToSet));
+          int counter = 0;
+          for (List<ColumnStatistics> sublist : Lists.partition(statisticsToSet, partitionBatchSize)) {
+            int start = counter * partitionBatchSize;
+            LOG.info("Setting column statistics for partitions {} through {}", start, start + sublist.size() - 1);
+            client.setPartitionColumnStatistics(new SetPartitionsStatsRequest(statisticsToSet));
```
Comment (abhimanyugupta07) [inline circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java]: here too. `sublist` instead of `statisticsToSet` ?

File: circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java [hunk 3]
```diff
@@ -425,18 +440,28 @@ public void alteringExistingPartitionedReplicaTableWithPartitionsSucceeds() thro
...
+    // we sent a list of 3 new partitions and used a partitionBatchSize of 2
+
+    List<List<Partition>> addCaptorValues = addPartitionCaptor.getAllValues();
+    List<Partition> firstBatch = addCaptorValues.get(0);
+    List<Partition> secondBatch = addCaptorValues.get(1);
+    assertThat(addCaptorValues.size(), is(2));
+    assertThat(firstBatch.size(), is(2));
+    assertThat(secondBatch.size(), is(1));
 
     Partition altered = alterPartitionCaptor.getValue().get(0);
```
Comment (abhimanyugupta07) [inline circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java]: I think to test the batching for `alter_partition` and `setPartitionColumnStatistics` usecase, it will be nice to have more than one existing partitions.

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java [hunk 4]
```diff
@@ -30,6 +30,8 @@
 import java.util.Map;
 import java.util.Objects;
 
+import com.google.common.annotations.VisibleForTesting;
```
Comment (massdosage) [inline circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java]: Minor, but these imports should go along with the other google imports. I've made a mental note to update our hotels-oss-parent with a plugin that automatically does this that we've started using in EG org on GitHub so this is automated by Maven.

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java [hunk 5]
```diff
@@ -89,6 +93,29 @@
     this.housekeepingListener = housekeepingListener;
     replicationMode = tableReplication.getReplicationMode();
     this.tableReplication = tableReplication;
+    this.partitionBatchSize = DEFAULT_PARTITION_BATCH_SIZE;
```
Comment (massdosage) [inline circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java]: NABD but you could inline this on line 77 and then only have the assignment in the constructor below which actually allows you to override the value.

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java [hunk 6]
```diff
@@ -89,6 +93,29 @@
...
+  }
+
+  /**
+   * Use {@link ReplicaFactory}
```
Comment (massdosage) [inline circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java]: What does this comment mean?

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java [hunk 7]
```diff
@@ -175,7 +202,13 @@ public void updateMetadata(
       if (!partitionsToCreate.isEmpty()) {
         LOG.info("Creating {} new partitions.", partitionsToCreate.size());
         try {
-          client.add_partitions(partitionsToCreate);
+          int counter = 0;
+          for (List<Partition> sublist : Lists.partition(partitionsToCreate, partitionBatchSize)) {
```
Comment (massdosage) [inline circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java]: I've never come across `Lists.partition` before, looks good.

File: circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java [hunk 8]
```diff
@@ -22,22 +22,29 @@
...
-import java.io.IOException;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.Comparator;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-
+import com.google.common.base.Supplier;
```
Comment (massdosage) [inline circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java]: I know it's not the most readable format but our import ordering is this: https://github.com/ExpediaGroup/eg-oss-parent/blob/master/pom.xml#L102

File: circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java (hunk 3 above)
Comment (massdosage) [reply circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java]: Yeah, it feels like we are missing some tests that would have caught the issues you called out above. I'd suggest writing new tests (or modifying existing) and getting them to fail with the current code and then fixing the code and they should then pass.

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java (hunk 1 above)
Comment (barnharts4) [reply circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java]: good catch, thanks!

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java (hunk 2 above)
Comment (barnharts4) [reply circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java]: yep, thanks

File: circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java (hunk 3 above)
Comment (barnharts4) [reply circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java]: hmm, I agree with adding more tests for for `alter_partition` and `setPartitionColumnStatistics` but how would you suggest I get current(old) code to fail?  I guess I could mock a failure of the metastore client if the size of the list was longer than the `partitionBatchSize`, but that seems fairly contrived and not really testing anything except my ability to mock stuff.

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java (hunk 4 above)
Comment (barnharts4) [reply circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java]: Yeah, I am struggling with the correct IntelliJ settings to match circus train's preferred import layout.  Obviously DSP standards are not matching up well with HCOM in this circumstance. Do you have a doc or a settings file or something so that I can update?

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java (hunk 6 above)
Comment (barnharts4) [reply circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java]: It was copied from the other constructor, and it means that this constructor is package-private and to construct this, you need to use `ReplicaFactory`.

File: circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java (hunk 8 above)
Comment (barnharts4) [reply circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java]: ah, thanks, just what I was looking for!

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java (hunk 4 above)
Comment (massdosage) [reply circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java]: This will do it once I've merged and released the parent - https://github.com/HotelsDotCom/circus-train/pull/168/files

I'm getting timeouts publishing to Maven Central right now, they've had lots of scalability issues in the last few months so I'm going to leave it for today and try again tomorrow.

File: circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java (hunk 3 above)
Comment (barnharts4) [reply circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java]: or maybe you didn't mean the issue of overwhelming the HMS with the full list, but rather the issue that I forgot to use `sublist` above?  In that case, yes, good point.

Comment (abhimanyugupta07) [reply circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java]: Yes. Your last comment is what I meant. Just adding two more existing partitions in the test should give you a batch to trigger the changes you made.

**Comment by barnharts4** [comment]
Updated test to work with any number of added and altered partitions, and any size batch.  Added tests for common edge cases and boundary conditions.  Addressed PR comments.

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java (hunk 6 above)
Comment (massdosage) [reply circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/Replica.java]: Ah, OK.

File: circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java [hunk 9]
```diff
@@ -423,41 +466,106 @@ public void alteringExistingPartitionedReplicaTableWithPartitionsSucceeds() thro
...
+      assertThat(columnStats.get(colStatNdx).getStatsDesc().getPartName(),
+              is(String.format("c=new_%s/d=new_%s_sub", addPartColStatNdx, addPartColStatNdx)));
+      assertThat(columnStats.get(colStatNdx).getStatsObj().size(), is(2));
+    }
+  }
+
+
+  @Test
+  public void alteringExistingPartitionedReplicaTableWithNewPartitionsInBatchesSucceeds_0_0() throws TException, IOException {
+    alterExistingPartitionedReplicaTableWithNewPartitionsInBatches(0,0);
```
Comment (massdosage) [inline circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java]: For another day but I wonder if we could use JUnit's "parameterised" feature to do this.

Comment (barnharts4) [reply circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java]: I looked into that, but found referenced to Parameterized not working with `MockitoJUnitRunner` (https://stackoverflow.com/questions/12606148/passing-parameterized-input-using-mockitos) so I didn't pursue further in the interest of time.  But yes, that would be the perfect for this situation.

Comment (massdosage) [reply circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/ReplicaTest.java]: Ah, OK. We should move to JUnit5 at some point as then it can definitely be done, see https://github.com/ExpediaGroup/beekeeper/blob/master/beekeeper-path-scheduler-apiary/src/test/java/com/expediagroup/beekeeper/scheduler/filter/WhitelistedListenerEventFilterTest.java
